space-invaders.asm

```diff
@@ -1,12 +1,65 @@
 ; constants
-%define NUM_INVADERS 12
+
+; GAME WINDOW PARAMETERS
 %define GAME_WIDTH 27
 %define GAME_HEIGHT 25
 
+; GAME PARAMETERES
+%define NUM_INVADERS 12
 %define INVADERS_MOVE_CYCLES 40
 %define INVADERS_SHOOT_CYCLES 4
 %define BULLETS_MOVE_CYCLE 28
 
+; SPECIAL CONSTANT
+%define INVALID_STATE 0x0000
+
+; ICONS
+%define PLAYER 'M'
+%define INVADERS 'T'
+%define BULLET '|'
+%define EXPLOSION_BULLET '#'
+%define WALL '#'
+
+; GAME WINNERS
+%define PLAYER_WIN 0
+%define INVADERS_WIN 1
+
+; GAME STATES
+%define START_STATE 0
+%define GAME_STATE 1
+%define END_STATE 2
+
+; PLAY KEYS
+%define START_KEY ' '
+%define RETRY_KEY 'r'
+%define MOVE_LEFT_KEY 'a'
+%define MOVE_RIGHT_KEY 'd'
+%define SHOOT_KEY ' '
+
+; MOVE DIRECTIONS
+%define MOVE_UP 0
+%define MOVE_RIGHT 1
+%define MOVE_DOWN 2
+%define MOVE_LEFT 3
+%define MOVE_RESET 4
+
+; BULLET
+%define BULLET_STATUS_END_OF_LIST 0
+%define BULLET_STATUS_EXPLOSION '#'
+%define BULLET_STATUS_PLAYER 'p'
+%define BULLET_STATUS_INVADER 'i'
+
+%define BULLET_STATUS_OFFSET 0
+%define BULLET_STATUS_SIZE 1
+%define BULLET_POSITION_OFFSET BULLET_STATUS_OFFSET + BULLET_STATUS_SIZE
+%define BULLET_POSITION_SIZE 2
+%define BULLET_SIZE BULLET_POSITION_OFFSET + BULLET_POSITION_SIZE
+
+; INVADERS
+%define INVADERS_POSITION_OFFSET 0
```

Comment:
Very minor remark, but I would prefer to call this `INVADER` instead of `INVADERS`, since one would also declare an invader struct using a singular term.

Reply:
Oh, I see.
```BULLET``` is also singular.

Reply:
```
 ; invaders
  invaders resw NUM_INVADERS
  num_invaders_alive resb 1
  invaders_move_direction resb 1
  invaders_move_cycle resb 1
  invaders_shoot_cycle resb 1
```

Regarding the above, Can I keep the plural form of the variable name? Should I correct it to singular form?